lib/events/codes.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package events
+
+import (
+	"sync"
+
+	"github.com/gravitational/trace"
+)
+
+// init registers all Teleport events.
+func init() {
+	for _, event := range All {
+		if err := RegisterEvent(event); err != nil {
+			panic(err)
+		}
+	}
+}
+
+// eventsRegistry contains all registered in the system events.
+var eventsRegistry = make(map[string]EventInfo)
+
+// m is a mutex protecting events registry.
+var m sync.RWMutex
+
+// RegisterEvent adds the specified event to the global registry.
+func RegisterEvent(event EventInfo) error {
```

Comment:
can you look for a way around having a global concurrently accessed map? AFAIK the sole purpose of this map is to  augment the missing fields using LookupEvent in add additional fields function.

What if instead, you would make `Code` type  with the `AddFields` 

```go
type Code string
func (c Code) AddFields(event.Fields) {
   ..
}

```

Or may be figure out something else to avoid this global lookup/register pattern

Reply:
@klizhentas Actually, the primary reason I took this approach is not just to augment the fields, but so it is easy to register new types of events - for example Gravity would just "register" its own event types and this way audit log will automatically be able to augment their fields too. It's kinda similar to the "register" functions used in other places in Teleport - e.g. IIRC there is "register xxx resource marshaler/unmarshaler" for various resources for the same purpose - so it's easy to extend the implementation.

I was also considering another approach - to have a some kind of interface (`type Augmentor interface {}`) which Gravity would extend - but I thought the downside of that approach is that it would have to be initialized and passed it everywhere to all audit log implementations so it'd be more verbose / require more boilerplate.

Reply:
what's the purpose of registering an event?

Reply:
Basically, the purpose of registering a particular event is to let audit log know how to augment its fields. It would not be needed if we only had Teleport but we also have Gravity so audit log needs to know how to update Gravity-specific events too.

Reply:
I think it's backwards - every event should know how to augment it's fields, not vice versa - if you do it the other way around, there won't be any need to create a global map

Reply:
I agree that event should be in charge of it's fields. Initially I was thinking about something like this:

```
const (    
    ForwarderDeleted = NewWarningEvent({
        Code: "G2001W",
        CodeDescription: "Log Forwarder Deleted"        
        Message: "New Log Forwarder {{ .Name }} has been created"
    })
)

// usage
eventInstance := ForwarderDeleted.NewInstance(fields) // random property bag 
libevents.Emit(ctx, eventInstance)
```

Reply:
I suggest to drop this field for now